homeassistant/components/sonos/sensor.py

```diff
@@ -0,0 +1,182 @@
+"""Entity representing a Sonos Move battery level."""
+import logging
+from typing import Any, Dict, Optional
+
+from pysonos.core import SoCo
+from pysonos.exceptions import SoCoException
+
+from homeassistant.const import DEVICE_CLASS_BATTERY, PERCENTAGE, STATE_UNKNOWN
+from homeassistant.core import Event, callback
+import homeassistant.helpers.device_registry as dr
+from homeassistant.helpers.entity import Entity
+from homeassistant.helpers.icon import icon_for_battery_level
+
+from .const import (
+    DATA_SONOS,
+    DOMAIN as SONOS_DOMAIN,
+    SCAN_INTERVAL,
+    SONOS_DISCOVERY_UPDATE,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+ATTR_BATTERY_LEVEL = "battery_level"
+ATTR_BATTERY_CHARGING = "charging"
+ATTR_BATTERY_POWERSOURCE = "power_source"
+
+
+async def async_setup_entry(hass, config_entry, async_add_entities):
+    """Set up Sonos from a config entry."""
+
+    def _async_create_entities(event: Event):
+        soco = event.data.get("soco")
+        if soco and soco.uid not in hass.data[DATA_SONOS].battery_entities:
+            hass.data[DATA_SONOS].battery_entities[soco.uid] = None
+            hass.async_add_executor_job(_discover_battery, hass, soco)
+
+    def _discover_battery(hass, soco):
+        battery_info = SonosBatteryEntity.fetch(soco)
+        if battery_info is not None:
+            hass.add_job(async_add_entities, [SonosBatteryEntity(soco, battery_info)])
```

Comment:
I think we can simplify this a bit by collecting the battery info before we jump back from the executor then we can do all this in the event loop

Reply:
I took care of this